core/modules/replica/DatabaseMySQLTypes.h

```diff
@@ -0,0 +1,216 @@
+ *   port: 3306
+ *   user: effective user of a process
+ *
+ * The following parameters will be empty:
+ *
+ *   password:
+ *   database:
+ */
+struct ConnectionParams {
+
+    /// The DNS name or IP address of a machine where the database
+    /// server runs
+    std::string host;
+
+    /// The port number of the MySQL service
+    uint16_t port;
+
+    /// The name of a database user
+    std::string user;
+
+    /// The database password
+    std::string password;
+
+    /// The name of a database to be set upon the connection
+    std::string database;
+
+    /**
+     * The factory method will return an instance of this structure initialized
+     * by values of parameters found in the input encoded string. The string is
+     * expected to have the following syntax:
+     *
+     *   mysql://[user][:password]@[host][:port][/database]
+     *
+     * NOTES ON THE SYNTAX:
+     * 1) all keywords are madatory
+     * 2) the corresponding values for for all but the database are optional
+     * 3) default values for other parameters (if missing in the string) will be assumed.
+     *
+     * @param params          - connection parameters packed into a string
+     * @param defaultHost     - default value for a host name
+     * @param defaultPort     - default port number
+     * @param defaultUser     - default value for a database user account
+     * @param defaultPassword - default value for a database user account
+     *
+     * @throw std::invalid_argument - if the string can't be parsed
+     */
+    static ConnectionParams parse(std::string const& params,
+                                  std::string const& defaultHost,
+                                  uint16_t defaultPort,
+                                  std::string const& defaultUser,
+                                  std::string const& defaultPassword);
+
+    /// Default constructor will initialize connection parameters
+    /// with default values
+    ConnectionParams();
+
+    /**
+     * Return a string representation of all (but the password) parameters.
+     * The result will be formatted similarily to the one expected by
+     * the non-default constructor of the class.
+     */
+    std::string toString() const;
+};
+
+/// Overloaded operator for serializing ConnectionParams instances
+std::ostream& operator<<(std::ostream&, ConnectionParams const&);
+
+/**
+ * Class DoNotProcess is an abstraction for SQL strings which than ordinary
+ * values of string types needs to be injected into SQL statements without
+ * being processed (excaped and quoted) as regular string values.
+ */
+class DoNotProcess {
+
+public:
+
+    /**
+     * The normal constructor
+     *
+     * @param name_ - the input value
+     */
+    explicit DoNotProcess(std::string const& name_);
+
+    DoNotProcess() = delete;
+
+    DoNotProcess(DoNotProcess const&) = default;
+    DoNotProcess& operator=(DoNotProcess const&) = default;
+
+    virtual ~DoNotProcess() = default;
```

Comment:
I'm not sure why does it need to be virtual at all, there are no other methods defined in this class. To me it looks like you are using sub-classing as a factory to create different values of this class. Maybe just regular factory function be enough for that.

Reply:
May be I don't need to have it virtual. I put it here simply because this is a base class for a few others. A family of these types is used to steer proper SQL generation in header DatabaseMySQL.h